Duplicati/Library/AutoUpdater/DataFolderLocator.cs

```diff
@@ -0,0 +1,133 @@
+    /// <returns>The default storage folder</returns>
+    public static string GetDefaultStorageFolder(string targetfilename)
+        => DataFolderManager.OVERRIDEN_DATAFOLDER
+            ? DataFolderManager.DATAFOLDER
+            : GetDefaultStorageFolderInternal(targetfilename, AutoUpdateSettings.AppName);
+
+    /// <summary>
+    /// Finds a default storage folder, using the operating system specific locations.
+    /// The targetfilename is used to detect locations that are used in previous versions.
+    /// If the targetfilename is found in an old location, but not the current, the old location is used.
+    /// Note that the folder is not created, only the path is returned.
+    /// </summary>
+    /// <param name="targetfilename">The filename to look for</param>
+    /// <param name="appName">The name of the application</param>
+    /// <returns>The default storage folder</returns>
+    internal static string GetDefaultStorageFolderInternal(string targetfilename, string appName)
+    {
+        //Normal mode uses the systems "(Local) Application Data" folder
+        // %LOCALAPPDATA% on Windows, ~/.config on Linux
+        var folder = System.IO.Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), appName);
+
+        if (OperatingSystem.IsWindows())
+        {
+            // Special handling for Windows:
+            //   - Older versions use %APPDATA%
+            //   - New versions use %LOCALAPPDATA%
+            //   - And prevent using C:\Windows\
+            var newlocation = System.IO.Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), appName);
+
+            var folderOrder = new List<string>() {
+                    folder,
+                    newlocation
+                };
+
+            // If %LOCALAPPDATA% is inside the Windows folder, prefer a LocalService folder instead
+            if (Common.IO.Util.IsPathUnderWindowsFolder(newlocation))
+            {
+                var userProfilesFolder = Library.Utility.SHGetFolder.UserProfilesFolder;
+                if (!string.IsNullOrWhiteSpace(userProfilesFolder))
+                    folderOrder.Add(System.IO.Path.Combine(userProfilesFolder, "LocalService", appName));
+            }
+
+            // Prefer the most recent location
+            var matches = folderOrder.AsEnumerable()
+                .Reverse()
+                .Where(x => System.IO.File.Exists(System.IO.Path.Combine(x, targetfilename)))
+                .ToList();
+
+            // Use the most recent location found with content
+            // If none are found, use the most recent location
+            folder = matches.FirstOrDefault() ?? folderOrder.Last();
+        }
+
+        if (OperatingSystem.IsMacOS())
+        {
+            // Special handling for MacOS:
+            //   - Older versions use ~/.config/
+            //   - but new versions use ~/Library/Application\ Support/
```

Comment:
The PR text states that "For MacOS there are no changes.", but this comment makes it sound like it has changed? Or is the "old/new" relating to a previous folder change?

Reply:
The MacOS placement changed with the switch from Mono to .NET8. Due to refactoring the `DatabaseLocator.cs` into two files this shows up as "new" code, but it has not been changed.